src/coreclr/System.Private.CoreLib/src/System/Runtime/DependentHandle.cs

```diff
@@ -0,0 +1,130 @@
+        // merely wraps an IntPtr so this struct serves mainly as a "managed typedef."
+        //
+        // DependentHandles exist in one of two states:
+        //
+        //    IsAllocated == false
+        //        No actual handle is allocated underneath. Illegal to get Target, Dependent
+        //        or GetTargetAndDependent(). Ok to call Dispose().
+        //
+        //        Initializing a DependentHandle using the nullary ctor creates a DependentHandle
+        //        that's in the !IsAllocated state.
+        //        (! Right now, we get this guarantee for free because (IntPtr)0 == NULL unmanaged handle.
+        //         ! If that assertion ever becomes false, we'll have to add an _isAllocated field
+        //         ! to compensate.)
+        //
+        //
+        //    IsAllocated == true
+        //        There's a handle allocated underneath. You must call Dispose() on this eventually
+        //        or you cause a native handle table leak.
+        //
+        // This struct intentionally does no self-synchronization. It's up to the caller to
+        // to use DependentHandles in a thread-safe way.
+        // =========================================================================================
+
+        private IntPtr _handle;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DependentHandle"/> struct with the specified arguments.
+        /// </summary>
+        /// <param name="target">The target object instance to track.</param>
+        /// <param name="dependent">The dependent object instance to associate with <paramref name="target"/>.</param>
+        public DependentHandle(object? target, object? dependent)
+        {
+            // no need to check for null result: nInitialize expected to throw OOM.
+            _handle = nInitialize(target, dependent);
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether this handle has been allocated or not.
+        /// </summary>
+        public bool IsAllocated => _handle != IntPtr.Zero;
+
+        /// <summary>
+        /// Gets or sets the target object instance for the current handle.
+        /// </summary>
+        public object? Target
+        {
+            get => nGetPrimary(_handle);
+            set => nSetPrimary(_handle, value);
+        }
+
+        /// <summary>
+        /// Gets or sets the dependent object instance for the current handle.
+        /// </summary>
+        /// <remarks>
+        /// If it is necessary to retrieve both <see cref="Target"/> and <see cref="Dependent"/>, it is
+        /// recommended to use <see cref="GetTargetAndDependent"/> instead. This will result in better
+        /// performance and it will reduce the chance of unexpected behavior in some cases.
+        /// </remarks>
+        public object? Dependent
+        {
+            get => GetTargetAndDependent().Dependent;
```

Comment:
Do we need to add a dedicated method to make this efficient?

Reply:
I think that might not be needed in this case, for two reasons:
1) The comments mention that (on CoreCLR) retrieving the dependent handle is the most expensive operation to ask the GC to do, so I'd expect that in this scenario that cost would probably dwarf the extra register shuffle here to discard the target and return the dependent. So don't think having a dedicated internal call just to expose a better signature would be worth the effort.
2) I'd expect most devs to either just check the target, or just to call `GetTargetAndDependent()` directly (like `CWT<,>` does as well). This `Dependent` getter was mostly added for completeness to the public API, but not sure it's worth optimizing too much unless maybe someone points out how that's an issue?

Does this make sense? 🙂